[Feat] #56 - 푸시알림 세팅 뷰 Domain, Data, Network 레이어 연결

yangsubinn opened the pull request:
### 🛠 작업 내용
<!-- 작업한 내용을 적어주세요. -->
- 공지리스트 뷰에서 설정 화면 전환 연결
- 설정 뷰, 푸시알림 설정 뷰 네비바 추가
- PushAlarmSettingEntity, Model, Transform 구현
- AlertService, API에 fetchPushSettingList 추가
- PushAlarmSettingRepository, useCase 구현
- PushAlarmSettingVC, ViewModel 바인딩
- PartCategory enum형 core로 분리

### 🚨 참고 사항
<!-- 참고할 사항이 있다면 적어주세요. -->
#### ✅`PartCategory`라는 enum형을 `Core` 모듈에 넣어뒀습니다
<img width="587" alt="스크린샷 2022-10-27 오후 1 02 28" src="https://user-images.githubusercontent.com/81167570/198188374-04a07940-b00b-4286-85ca-10b028cbbb1e.png">

- `PushAlarmSettingEntity`에서 part 부분을 받아올때 해당 enum 형을 사용하려고 했슴니다    
  * 해당 enum형을 공통으로 사용할 수 있을 것 같았는데, 그러다보니 Network외의 Presentation이나 Domain에서eh 모두 접근 가능해야해서 일단 Core의 Literals에 넣어뒀습니다 
- 근데 Literals이라는 폴더에 들어가는게 조금 어색한가 싶긴 해여 .. 좋은 의견 있으면 남겨주세요🙇‍♀️

<img width="600" alt="스크린샷 2022-10-27 오후 1 07 48" src="https://user-images.githubusercontent.com/81167570/198189052-2485f40d-4f94-4932-a975-24e562dcebc4.png">
<img width="600" alt="스크린샷 2022-10-27 오후 1 08 45" src="https://user-images.githubusercontent.com/81167570/198189183-4401c583-963e-4aa5-8a29-6926aa8ad388.png">

- ViewModel 안에 있던 PartList를 주석처리하고 PartCategory의 extension을 사용해서 수정해뒀습니다

- 코드리뷰 후 괜찮으면 주석 부분 삭제해서 머지하겠습니다 !


### 📸 스크린샷
|기능|스크린샷|
|:--:|:--:|
| 푸시알림 설정 뷰 | <img src="https://user-images.githubusercontent.com/81167570/198189875-58ddc9fc-c8fc-422b-aea2-2fd5995796a6.jpeg" width=300> |


### 💡 관련 이슈
<!-- 관련있는 이슈 번호(#000)를 적어주세요. -->
closed #56

yangsubinn committed:
[Feat] #56 - 공지리스트뷰에서 설정 화면 전환, 설정 뷰 네비 추가

yangsubinn committed:
[Feat] #56 - 푸시알림 세팅 뷰 네비바 추가

yangsubinn committed:
[Merge] develop pull
Merge branch 'develop' of https://github.com/sopt-makers/SOPT-iOS into feature/#56-SettingPushalarmList

yangsubinn committed:
[Feat] #56 - PushAlarmSetting Transform, Modle, Entity 구현

yangsubinn committed:
[Feat] #56 - PushAlarmSetting Network 모듈 구현

yangsubinn committed:
[Feat] #56 - PushAlarmSetting repository, usecase 구현

yangsubinn committed:
[Fix] #56 - PushAlarmSettingModel, Transform 변형

yangsubinn committed:
[Feat] #56 - PushAlarmSetting Presentation 모듈 구현

yangsubinn committed:
[Fix] #56 - PartCategory Core 모듈로 분리, 적용

yangsubinn committed:
[Chore] #56 - PartCategory extension 위치 수정

L-j-h-c commented on SOPT-iOS/Projects/Core/Sources/Literals/PartCategory.swift:
```diff
@@ -0,0 +1,42 @@
+
+import Foundation
+
+public enum PartCategory: String, Codable {
+    case fullNotice = "all"
+    case plan = "plan"
+    case design = "design"
+    case ios = "iOS"
+    case android = "android"
+    case web = "web"
+    case server = "server"
+}
+
+extension PartCategory {
+    public static var allCases: [PartCategory] = [.fullNotice, .plan, .design, .ios, .android, .web, .server]
+    
+    public var title: String {
+        switch self {
+        case .fullNotice:
+            return "전체 공지"
+        case .plan:
+            return "기획"
+        case .design:
+            return "디자인"
+        case .ios:
+            return "iOS"
+        case .android:
+            return "Android"
+        case .web:
+            return "Web"
+        case .server:
+            return "Server"
+        }
+    }
+}
```
👍 👍 
깔끔한 모델 감사합니다. 개인적으로 파트 카테고리에 관한 정보들은 앱의 핵심 로직 또는 모델에 해당하기 때문에 domain 안에 둬도 좋을 것 같은데 어떨까요? import Domain을 해야하는 부분이 번거롭기는 하지만...하하

lsj8706 replied:
제 생각에도 해당 enum에 나중에도 혹시라도 추가적인 함수(기능)나 변수가 포함 될 가능성이 있기 때문에 Literals도 좋지만 Domain쪽이 더 적합해 보이기는 합니다...!

yangsubinn replied:
제가 그 이 친구를 Core에 넣어뒀던 이유가 요 아래 Entity에서도 사용 중이라 Domain에 넣으면 접근할 수가 없어서 그랬었습니다 .. !
<img width="663" alt="스크린샷 2022-10-28 오전 1 04 10" src="https://user-images.githubusercontent.com/81167570/198341369-0467ef05-9d83-4f5d-bcec-ae9239c74c38.png">

만약 함수나 기능의 확장성을 위해서라면, Core에는 기본 enum형만 두고 Domain에서 extension으로 빼서 사용할 수 있을 것 같긴 한디
파트 카테고리 관한 정보가 핵심 로직에 포함되니까 그래두 Domain으로 옮기는게 나을 것 같으면, 
고냥 저 자체는 Codable 빼고 Domain으로 옮기구 Entity에서는 따로 enum형 만들어서 사용할게여 !

**❓Domain으로 옮기고 Entity용은 따로 만들어서 쓸까요?** @L-j-h-c @lsj8706 @devxsby

L-j-h-c replied:
오오...개인적으로 entity는 서버에서 오는 response를 그대로 받아오는 것이라고 할 떄, String으로 part를 받아온 다음에 Data Module에서 toDomain을 하는 과정에 import Domain해서 PartCategory로 바꿔주는게 맞는 흐름이라고 생각했습니다...
근데 enum형이 codable하게 저렇게 될 수 있는지 몰랐네요...

yangsubinn replied:
훔 response 그대로 받아오는거니까 그렇긴 하네여 .. 
그럼 일단 entity에서는 string으로 받아오도록 하고 PartCategory는 domain으로 옮기는 방향으로 수정해서 다시 언급 한번 드리겠습니다 !

yangsubinn replied:
요거는 일단 수정했습니다 !

L-j-h-c commented on SOPT-iOS/Projects/Presentation/Sources/SettingScene/VC/PushAlarmSettingVC.swift:
```diff
@@ -20,9 +20,21 @@ public class PushAlarmSettingVC: UIViewController {
     
     public var viewModel: PushAlarmSettingViewModel!
     private var cancelBag = CancelBag()
+    private var pushToggleList = Array(repeating: false, count: 7)
```
MVVM에서 VM이 View의 State를 관리하고 있다고 생각했을 때, 이런 정보들은 viewModel쪽에서 저장하면 좋다고.. 개인적으로 생각하고 있기는 합니다. 이유는 MVVM에서 View를 최대한 dumb하게 만들어서, 서로 다른 ViewModel의 상태를 바꿔서 사용하는 대로 원하는 결과를 예상할 수 있게 되기 때문(Testable해진다)인 것 같다는... 뇌피셜입니다!
방법론적인 부분이라 이 원칙을 어느정도로 따라야 할지는 잘 판단해서 유동적으로 하면 되는 것 같습니다!

L-j-h-c replied:
근데 또 viewModel.[list].value 이런식으로 가져가는게 맞는 방식인지는 또 확신이 없어서... 계속 고민해보면 좋을 것 같아요 의견 부탁드립니다 ㅎㅎ

yangsubinn replied:
엄 output에서 받아온 값을 바로 cell에 적용해보려고 하다가 apply하거나 그럴 수 있는 형태가 아니라 그냥 에라이 하고 
pushToggleList에 값 저장하고 reload 해주는 식으로 했던 것 같네여 .. 

viewModel에서 여러 곳에서 공통으로 쓰이는? 일회성이 아닌 데이터는 저장해두고 가져다 쓰는게 맞다구 생각하는데
이게 usecase에서 받아온 값을 bindOutput에서 바로 넘겨받아서 온 값이라 흠
이걸 또 viewModel에 저장하고 다시 output이 아닌 viewModel에 접근해서 사용하는게 맞을지 헷갈리네여..

yangsubinn replied:
vc에서 저장 안하고 해보려구 했는데, viewModel에서 output으로 나온 값을 저장했다가 다시 viewModel에서 값을 가져다가 쓰는 방법밖에 생각이 안나더라구요 .. 근데 요 방법은 좀 output을 제대로 활용하지 못하는 방법인 것 같아서 일단은 이렇게 두겠습니다 ..!

추가로,, pushToggleList에서 초기값을 넣어둔 이유는 
tableview cellForRowAt에서 pushToggleList[indexPath.item]으로 넣어주고 있었는데
빈 배열로 두면 index 에러가 떠서 false로 넣어뒀습니다 (cell의 개수는 PartCategory 개수 자체로 정해주고 있어서)

L-j-h-c reviewed (approved):
모델 감사합니다... 👼 👼 수고하셨어용~!

lsj8706 commented on SOPT-iOS/Projects/Presentation/Sources/PostListScene/VC/PostListVC.swift:
```diff
@@ -34,7 +33,8 @@ public class PostListVC: UIViewController {
         .setTitle(I18N.Navi.post)
         .rightButtonAction {
-            print("rightButtonAction")
+            let settingVC = self.factory.makeSettingVC()
```
👍👍

lsj8706 reviewed (approved):
수고하셨습니다~!!

yangsubinn committed:
[Merge] develop pull
Merge branch 'develop' of https://github.com/sopt-makers/SOPT-iOS into feature/#56-SettingPushalarmList

yangsubinn committed:
[Merge] develop pull

yangsubinn committed:
[Move] #56 - PartCategory Domain으로 이동

yangsubinn committed:
[Fix] #56 - PartCategory iOS에서 ios로 수정
